### PR TITLE
Handling nested predicates with relative references

### PR DESCRIPTION
**Pull request by phillipm**
An unfortunately large-ish PR that accomplishes the following
 - make sure (relative) predicates in triggers are contextualized correctly
 - parser warnings for control references that have predicates
 - parser warnings for binds/control elements with references pointing to non-main instances
 - cleanup of EvaluationContext.expandReferenceAccumulator
 - some tests for these changes
 - misc comments and small code refactors.

Trying to address this [ticket](http://manage.dimagi.com/default.asp?89140)

### CODE REVIEW

File: core/src/org/javarosa/xpath/expr/XPathPathExpr.java [hunk 1]
```diff
@@ -170,7 +177,8 @@ public TreeReference getReference(boolean allowPredicates) throws XPathUnsupport
                 for (int j = 0; j < step.predicates.length; j++) {
                     v.addElement(step.predicates[j]);
                 }
-                ref.addPredicate(i, v);
```
Comment (phillipm) [inline core/src/org/javarosa/xpath/expr/XPathPathExpr.java]: If a (nested) predicate has ../, then "i" doesn't represent the correct step position to add the predicate to. This causes crashes on expressions like:
/data/group[count(../team[@id = 'mobile']) > 0])

File: core/src/org/javarosa/core/model/condition/EvaluationContext.java [hunk 2]
```diff
@@ -258,139 +260,141 @@ private void expandReferenceAccumulator(TreeReference sourceRef, DataInstance so
...
-                    }
-                    if (predicateEvaluationProgress != null) {
-                        predicateEvaluationProgress[0]++;
+    /**
+     * Load viable references that are children of the element argument.
+     *
+     * @param node element of which to collect child references
+     * @param name
+     * @param mult
+     * @param includeTemplates
```
Comment (phillipm) [inline core/src/org/javarosa/core/model/condition/EvaluationContext.java]: TODO

File: core/src/org/javarosa/form/api/FormEntryModel.java [hunk 3]
```diff
@@ -397,7 +397,7 @@ public boolean isIndexRelevant(FormIndex index) {
             relevant = form.isRepeatRelevant(ref);
         } else {
             TreeElement node = form.getMainInstance().resolveReference(ref);
-            relevant = node.isRelevant(); // check instance flag first
+            relevant = (node != null) && node.isRelevant();
```
Comment (phillipm) [inline core/src/org/javarosa/form/api/FormEntryModel.java]: Something is wrong with how heterogeneous paths are handled when checking relevancy (for <trigger> tags) that needs to be fixed. This just makes sure that if no node is found, we don't try to use it.

File: core/src/org/javarosa/xform/parse/XFormParser.java [hunk 4]
```diff
@@ -967,13 +967,25 @@ protected QuestionDef parseControl(IFormElement parent, Element e, int controlTy
         } else if (ref != null) {
             try {
                 dataRef = new XPathReference(ref);
+                TreeReference controlRefTarget = (TreeReference)dataRef.getReference();
+                if (controlRefTarget.getInstanceName() != null) {
+                    reporter.error("<" + e.getName() +
+                            "> points to an non-main instance (" +
+                            controlRefTarget.getInstanceName() +
+                            "), which isn't supported.");
```
Comment (phillipm) [inline core/src/org/javarosa/xform/parse/XFormParser.java]: warns against something like this
<trigger ref="instance('groups')/root/groups/group">

File: core/src/org/javarosa/xform/parse/XFormParser.java [hunk 5]
```diff
@@ -967,13 +967,25 @@ protected QuestionDef parseControl(IFormElement parent, Element e, int controlTy
...
+                TreeReference controlRefTarget = (TreeReference)dataRef.getReference();
+                if (controlRefTarget.getInstanceName() != null) {
+                    reporter.error("<" + e.getName() +
+                            "> points to an non-main instance (" +
+                            controlRefTarget.getInstanceName() +
+                            "), which isn't supported.");
+                }
+                if (controlRefTarget.hasPredicates()) {
+                    throw new XFormParseException("XForm Parse: The ref path " +
+                            "of a <trigger> isn't allowed to have predicates.", e);
```
Comment (phillipm) [inline core/src/org/javarosa/xform/parse/XFormParser.java]: Warns against something like \<trigger ref="/data/a/b/c[1]"\>

File: core/src/org/javarosa/xform/parse/XFormParser.java [hunk 6]
```diff
@@ -2449,8 +2461,20 @@ private void verifyBindings(FormInstance instance) {
                 i--;
             } else {
                 Vector<TreeReference> nodes = new EvaluationContext(instance).expandReference(ref, true);
+
                 if (nodes.size() == 0) {
-                    reporter.warning(XFormParserReporter.TYPE_ERROR_PRONE, "<bind> defined for a node that doesn't exist [" + ref.toString() + "]. The node's name was probably changed and the bind should be updated. ", null);
+                    if (ref.getInstanceName() != null) {
+                        reporter.warning(XFormParserReporter.TYPE_ERROR_PRONE,
```
Comment (phillipm) [inline core/src/org/javarosa/xform/parse/XFormParser.java]: warns against something like \<bind nodeset="instance('groups')/root/groups" ...\>

File: core/src/org/javarosa/xpath/XPathConditional.java [hunk 7]
```diff
@@ -120,24 +120,15 @@ private static void getExprsTriggersAccumulator(XPathExpression expr,
             TreeReference ref = ((XPathPathExpr)expr).getReference();
             TreeReference contextualized = ref;
 
-            if (ref.getContext() == TreeReference.CONTEXT_ORIGINAL) {
-                // Starts with 'current()' so contextualize in terms of the
-                // nodeset's original reference.
+            if (ref.getContext() == TreeReference.CONTEXT_ORIGINAL ||
+                    (contextRef == null && !ref.isAbsolute())) {
+                // Expr's ref begins with 'current()' or is relative and the
+                // context ref is missing.
```
Comment (phillipm) [inline core/src/org/javarosa/xpath/XPathConditional.java]: Allows for the relative path to be correctly contextualized in \<bind nodeset="/data/a/b/c" relevant="count(../../b[@id = 'two']) = 1" .../>

File: core/src/org/javarosa/core/model/condition/EvaluationContext.java [hunk 8]
```diff
@@ -230,18 +244,6 @@ public Object getVariable(String name) {
         return v;
     }
 
-    protected DataInstance retrieveInstance(TreeReference ref) {
```
Comment (phillipm) [inline core/src/org/javarosa/core/model/condition/EvaluationContext.java]: Just moving this down in the file

File: core/src/org/javarosa/core/model/condition/EvaluationContext.java [hunk 9]
```diff
@@ -259,141 +261,141 @@ private void expandReferenceAccumulator(TreeReference sourceRef, DataInstance so
...
+                    AbstractTreeElement child = node.getChild(name, i);
+                    if (child != null) {
+                        childSet.addElement(child.getRef());
+                    } else {
+                        throw new IllegalStateException("Missing or non-sequential nodes expanding a reference");
                     }
-                    if (passedAll) {
-                        expandReferenceAccumulator(sourceRef, sourceInstance, treeRef, refs, includeTemplates);
+                }
+                // TODO: this could be lifted out a conditional level, right? -- PLM
```
Comment (phillipm) [inline core/src/org/javarosa/core/model/condition/EvaluationContext.java]: Please double check, before I lift the "if (includeTemplates)" body out of the "if (mult == TreeReference.INDEX_UNBOUND)" body

File: core/src/org/javarosa/xpath/XPathConditional.java [hunk 10]
```diff
@@ -145,17 +136,26 @@ private static void getExprsTriggersAccumulator(XPathExpression expr,
...
-                //we can't generate this properly without an absolute reference
-                if (!ref.isAbsolute()) {
-                    throw new IllegalArgumentException("can't get triggers for relative references");
-                }
-                TreeReference predicateContext = ref.getSubReference(i);
+                // contextualizing with ../'s present means we need to
+                // calculate an offset to grab the appropriate predicates
+                int basePredIndex = contextualized.size() - ref.size();
+
+                TreeReference predicateContext = contextualized.getSubReference(basePredIndex + i);
```
Comment (phillipm) [inline core/src/org/javarosa/xpath/XPathConditional.java]: This makes sure (relative) predicates in triggers are contextualized

**Comment by benrudolph** [comment]
i love the comment updates. though future new hires will have an ever decreasing chance of stumbling across some real comment gems

File: core/src/org/javarosa/core/model/condition/EvaluationContext.java (hunk 9 above)
Comment (ctsims) [inline core/src/org/javarosa/core/model/condition/EvaluationContext.java]: Is this specific to this part of the code, I'm not quite seeing where this comment refers

Comment (phillipm) [inline core/src/org/javarosa/core/model/condition/EvaluationContext.java]: Lines 403 to 406. I could just leave it where-is, but it the template inclusion didn't seem to need to correspond to the fact that the childMult was INDEX_UNBOUND

**Pull request merged**